File: leetcode/problems/language-specific/go/50.pow-x-n.go
/*
 * @lc app=leetcode id=50 lang=golang
 *
 * [50] Pow(x, n)
 */

// @lc code=start
func myPow(x float64, n int) float64 {
	if n == 0 {
		return 1
	}
	if n < 0 {
		return myPow(1/x, -n)
	}
	if n%2 == 0 {
		return myPow(x*x, n/2)
	}

	return x * myPow(x*x, n/2)
}

// @lc code=end

File: leetcode/problems/sessions/2023-05-01_2023-??-??/50.pow-x-n.py
#
# @lc app=leetcode id=50 lang=python3
#
# [50] Pow(x, n)
#

# @lc code=start

# * Math Solution | O(logn) Time | O(1) Space
# * n -> The given input


class Solution:
    def myPow(self, x: float, n: int) -> float:
        if x == 0:
            return 0
        if n == 0:
            return 1

        if n < 0:
            return self.myPow(1 / x, -n)
        elif n % 2 == 0:
            return self.myPow(x * x, n // 2)

        return x * self.myPow(x * x, n // 2)


# @lc code=end
